test/unittests/evm_calls_test.cpp

```diff
@@ -741,6 +741,182 @@ TEST_P(evm, returndatacopy_outofrange)
     EXPECT_EQ(result.status_code, EVMC_SUCCESS);
 }
 
+TEST_P(evm, returndatacopy_outofrange_highbits)
+{
+    const uint8_t call_output[2]{};
+    host.call_result.output_data = std::data(call_output);
+    host.call_result.output_size = std::size(call_output);
+
+    // Covers an incorrect cast of RETURNDATALOAD arg to `size_t` ignoring the high bits.
+    const auto highbits =
+        0x1000000000000000000000000000000000000000000000000000000000000000_bytes32;
+    execute(735, staticcall(0) + returndatacopy(0, highbits, 0));
+    EXPECT_EQ(result.status_code, EVMC_INVALID_MEMORY_ACCESS);
+}
+
+TEST_P(evm, returndataload)
+{
+    // Not implemented in Advanced.
+    if (is_advanced())
+        return;
+
+    rev = EVMC_PRAGUE;
+    const auto call_output =
+        0x497f3c9f61479c1cfa53f0373d39d2bf4e5f73f71411da62f1d6b85c03a60735_bytes32;
+    host.call_result.output_data = std::data(call_output.bytes);
+    host.call_result.output_size = std::size(call_output.bytes);
+
+    const auto code = eof_bytecode(staticcall(0) + returndataload(0) + ret_top(), 6);
+    execute(code);
+    EXPECT_GAS_USED(EVMC_SUCCESS, 139);
+    EXPECT_EQ(bytes_view(result.output_data, result.output_size), call_output);
+}
+
+TEST_P(evm, returndataload_cost)
+{
+    // Not implemented in Advanced.
+    if (is_advanced())
+        return;
+
+    rev = EVMC_PRAGUE;
+    const uint8_t call_output[32]{};
+    host.call_result.output_data = std::data(call_output);
+    host.call_result.output_size = std::size(call_output);
+
+    const auto code = eof_bytecode(staticcall(0) + returndataload(0) + OP_STOP, 6);
+    execute(124, code);
+    EXPECT_EQ(result.status_code, EVMC_SUCCESS);
+    execute(123, code);
+    EXPECT_EQ(result.status_code, EVMC_OUT_OF_GAS);
+}
+
+TEST_P(evm, returndataload_outofrange)
+{
+    // Not implemented in Advanced.
+    if (is_advanced())
+        return;
+
+    rev = EVMC_PRAGUE;
+    {
+        const uint8_t call_output[31]{};
+        host.call_result.output_data = std::data(call_output);
+        host.call_result.output_size = std::size(call_output);
+
+        execute(124, eof_bytecode(staticcall(0) + returndataload(0) + OP_STOP, 6));
```

Comment:
Also I would suggest not to repeat `124` everywhere, just omit it, you tested it already in a separate test.

Reply:
I took it from an existing `returndatacopy` case. But in the case of `returndataload` I suppose it is more ok to drop and simplify